### PR TITLE
feat(highlight): allow extended highlighting patterns (#185)

### LINKED ISSUE #10: [Feature Request]: highlight author in `TODO(author):`
It's quite common for a TODO to have an author, I was using treesitter comment parser and it can highlight the author name inside the parentheses. Currently I use both to make this happen but it would be cool if this plugin can do that as well, thanks! :)

here's a reference:
image

### LINKED ISSUE #185: feature: TODO with numbers
### Did you check the docs?

- [X] I have read all the todo-comments.nvim docs

### Is your feature request related to a problem? Please describe.

no

### Describe the solution you'd like

I often use TODO labels with numbers, like `TODO 1 :` ` TODO 2 :`. I'm not sure whether the keywords support regular expressions, but if so I can do that with something like `TODO \d+ :` to match all TODO with numbers.
Besides, it'll be nice if the jumpings to next or previous TODO are numbers-aware. For example, I have many `TODO 2` and many `TODO 3`, and TODOs with the same number indicate some common goals, then I can jump to next `TODO 2` with something like `jump_next_same_number()`
And of course you can populate these things to other labels.

### Describe alternatives you've considered

no

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by LunarLambda**
Highlight the first capture group pattern, if there is a second (typically nested) capture group, use it for keyword matching.
Allows highlighting `TODO(foo)` using `.*<((KEYWORDS)%(\(.{-1,}\))?):`.

Closes #185, closes #10

This is similar to #180, however that PR's matching behaviour is not backwards compatible with the default highlight pattern used. This PR retains the behaviour that only the first capturing group is highlighted.

### CODE REVIEW

**Review by rubiin — APPROVED** [review]

**Comment by YuanYuYuan** [comment]
This PR is very helpful to me and doesn't seem to cause any obvious issues. Are there any concerns before merging?

**Comment by LunarLambda** [comment]
No concerns were raised here or in the original issue(s), the plugin maintainer just hasn't responded yet.

My PR branch tracks master, you could switch over to it for the time being.

**Comment by YuanYuYuan** [comment]
> My PR branch tracks master, you could switch over to it for the time being.

Hi @LunarLambda , yes, that's also my workaround right now. Thanks for your PR. And it's nice to have this merged. :smile: .

BTW, we may need to extend the search pattern as well. So I add the following pattern in my config

```lua
require("todo-comments").setup({
  highlight = {
    -- vimgrep regex, supporting the pattern TODO(name):
    pattern = [[.*<((KEYWORDS)%(\(.{-1,}\))?):]],
  },
  search = {
    -- ripgrep regex, supporting the pattern TODO(name):
    pattern = [[\b(KEYWORDS)(\(\w*\))*:]],
  }
})
```

**Comment by LunarLambda** [comment]
Rebased against latest master.

**Review by rubiin — APPROVED** [review]

**Comment by dpetka2001** [comment]
Would be nice to see this functionality in main plugin. Been using OP's branch without any apparent problems so far.

**Comment by LunarLambda** [comment]
It makes me happy that the PR is so well received. I will continue to rebase it against main since it's such a small change.

**Comment by sebszyller** [comment]
This has been waiting since March, and even Feb if you count #180. Any chance of merging so we can stop using OP's branch

**Comment by LunarLambda** [comment]
I've previously tried to notify folke once or twice in the issue where I originally proposed a workaround, which then became this PR.

I'm happy to keep rebasing my branch but I don't wish to bother folke about it.

**Comment by LunarLambda** [comment]
Apparently the linked issues will be closed by github actions soon due to being open for so long, as I just received multiple emails about it.

**Comment by folke** [comment]
Fixed by #180

**Pull request merged**
